File: java_solvers/src/main/java/edu/kit/ipd/java_solvers/ArithmeticEnumerationDemo.java
package edu.kit.ipd.java_solvers;

public class ArithmeticEnumerationDemo {

    private static long countOrModelsByEnumeration(int numVariables) {
        long result = 0;
        Assignment: for (long i = 0; i < Math.pow(2, numVariables); i++) {
            long remainingValue = i;
            for (int varIdx = 0; varIdx < numVariables; varIdx++) {
               if (remainingValue % 2 == 1) {
                   result += 1;
                   continue Assignment; // early abandoning: only one variable has to be true
               }
               remainingValue /= 2;
            }
        }
        return result;
    }

    // Slower version which uses built-in int-to-binary-conversion
    private static long countOrModelsByEnumeration2(int numVariables) {
        long result = 0;
        String iterationString = "";
        Assignment: for (long i = 0; i < Math.pow(2, numVariables); i++) {
            iterationString = Long.toBinaryString(i);
            for (int varIdx = 0; varIdx < numVariables; varIdx++) {
                if (varIdx < iterationString.length() && iterationString.charAt(varIdx) == '1') {
                    result += 1;
                    continue Assignment;
                }
            }
        }
        return result;
    }

    private static long countAndModelsByEnumeration(int numVariables) {
        long result = 0;
        Assignment: for (long i = 0; i < Math.pow(2, numVariables); i++) {
            long remainingValue = i;
            for (int varIdx = 0; varIdx < numVariables; varIdx++) {
               if (remainingValue % 2 == 0) {
                   continue Assignment; // early abandoning: all variables have to be true
               }
               remainingValue /= 2;
            }
            result += 1;
        }
        return result;
    }

    public static void main(String[] args) {
        long startTime = System.nanoTime();
        long solutionCount = countOrModelsByEnumeration(20);
        long endTime = System.nanoTime();
        System.out.println("OR - Time: " + (endTime - startTime) / 1e9 + " s, Solutions: " + solutionCount);

        startTime = System.nanoTime();
        solutionCount = countAndModelsByEnumeration(20);
        endTime = System.nanoTime();
        System.out.println("AND - Time: " + (endTime - startTime) / 1e9 + " s, Solutions: " + solutionCount);
    }

}

File: c_solvers/arithmetic_enumeration_demo.cpp
#include <chrono>
#include <iostream>

// use of unsigned types improves performance
unsigned long count_or_models_by_enumeration(unsigned short num_variables) {
    unsigned long result = 0;
    double iteration_count = std::pow(2, num_variables);
    for (unsigned long i = 0; i < iteration_count; i++) {
        unsigned long remaining_value = i;
        for (unsigned short var_idx = 0; var_idx < num_variables; var_idx++) {
            if (remaining_value % 2 == 1) {
                result += 1;
                break;
            }
            remaining_value /= 2;
        }
    }
    return result;
}

unsigned long count_and_models_by_enumeration(unsigned short num_variables) {
    unsigned long result = 0;
    double iteration_count = std::pow(2, num_variables);
    for (unsigned long i = 0; i < iteration_count; i++) {
        unsigned long remaining_value = i;
        for (unsigned short var_idx = 0; var_idx < num_variables; var_idx++) {
            if (remaining_value % 2 == 0) {
                goto Assignment_End;
            }
            remaining_value /= 2;
        }
        result += 1;
        Assignment_End:;
    }
    return result;
}

int main() {
    std::chrono::steady_clock::time_point startTime = std::chrono::steady_clock::now();
    long solutionCount = count_or_models_by_enumeration(30);
    std::chrono::steady_clock::time_point endTime = std::chrono::steady_clock::now();
    std::cout << "OR - Time: " << (endTime - startTime).count() / 1e9
        << " s, Solutions: " << solutionCount << "\n";

    startTime = std::chrono::steady_clock::now();
    solutionCount = count_and_models_by_enumeration(30);
    endTime = std::chrono::steady_clock::now();
    std::cout << "AND - Time: " << (endTime - startTime).count() / 1e9
        << " s, Solutions: " << solutionCount << "\n";
}
